DataScience/ImportantFeatures.py

```diff
@@ -0,0 +1,157 @@
+# sample: input => [['c','b','a','d','e'],['b','c','a'],['a']] returns output => [['a'], ['b', 'c'], ['d', 'e']]
+def get_feature_buckets(features_funnel):
+    union_features = []
+    feature_buckets = []
+    for features in reversed(features_funnel):
+        unique_features = list(set(features) - set(union_features))
+        unique_features.sort()
+        feature_buckets.append(unique_features)
+        union_features.extend(unique_features)
+    return feature_buckets
+
+def get_important_features(log_file, ml_args, warmstart_model=None, min_num_features=5):
+    invHash_fp = log_file+'.invHash.txt'
+    readModel_fp_base = log_file+'.readModel.{0}.txt'
+ 
+    if ' --l1 ' in ml_args: 
+        temp = ml_args.split(' --l1 ',1)
+        ml_args = temp[0]
+        if ' ' in temp[1]:
+            temp = temp[1].split(' ',1)
+            ml_args += ' ' + temp[1]
+            l1 = float(temp[0])
+        else:
+            l1 = float(temp[1])
+    else:
+        l1 = 1e-7
+    
+    vw_base = 'vw ' + ml_args + ' --dsjson --data {0} --quiet'.format(log_file)
+    if warmstart_model:
+        vw_base += ' -i {0}'.format(warmstart_model)
+
+    print('\n=====================================')
+    print('Generating invert hash file to map the hash to feature names')
+
+    # Use a low l1 value to capture all features
+    vw_inv_hash_cmd = vw_base + ' --invert_hash {0} --l1 {1}'.format(invHash_fp, 1e-23)
+    print('command to get invert hash file: ' + vw_inv_hash_cmd)
+    os.system(vw_inv_hash_cmd)
+    inv_hash = get_feature_inv_hash(invHash_fp)
+
+    print('\n=====================================')
+    print('Testing a range of L1 regularization')
+    all_features_funnel = []
+    index = 0
+    max_run_count = 20
+    while True:
+        readModel_fp = readModel_fp_base.format(index)
+        vw_readable_model_cmd_base = vw_base + ' --readable_model {0}'.format(readModel_fp)
+        vw_readable_model_cmd = vw_readable_model_cmd_base + ' -c --l1 {0}'.format(index)
+        index += 1
+        os.system(vw_readable_model_cmd)
+        features = extract_features(readModel_fp, inv_hash)
+        num_features = len(features)
+        print('L1: {0:.0e} - Num of Features: {1}, File - {2}'.format(l1, num_features, os.path.basename(readModel_fp)))
+        
+        all_features_funnel.append(features)
+        
+        # If we fall below the minimum number of features, then break out of the loop.
+        if num_features < min_num_features:
+            print('Number of features is {0} which is below the minimum of {1}. Exiting the loop with L1 value of: {2:.0e}'.format(num_features, min_num_features, l1))
+            break
+            
+        # Add a max run count so we avoid getting stuck in an infinite loop for any special case.
+        if index > max_run_count:
+            print('Run count exceeds max run count. Exiting the loop with L1 value of: {0:.0e}'.format(l1))
+            break
+        else:
+            l1 *= 10
```

Comment:
why are you multipling this if l1 is never used for your cmd: `vw_readable_model_cmd`

Reply:
Fixed it by using l1 instead of index in vw_readable_model_cmd.

Reply:
Fixed.

---
In reply to: [279973548](https://github.com/Microsoft/mwt-ds/pull/132#discussion_r279973548) [](ancestors = 279973548)